### ISSUE TITLE
dead code lint is incorrectly reporting code as being unused with static methods

### ISSUE THREAD

**Issue by erickt** [rust-lang/rust#19613]
I was playing with passing around a trait that only has static methods. It works okay, but it appears that this confuses the dead code lint. Here's an example:

``` rust
trait Visitor<R> {
    fn foo();
}

fn visit<V: Visitor<R>, R>() {
    Visitor::<R>::foo()
}

struct V;

impl Visitor<()> for V {
    fn foo() { println!("hi") }
}

pub fn main() {
    visit()
}
```

Which warns with:

```
serde2.rs:9:1: 9:10 warning: struct is never used: `V`, #[warn(dead_code)] on by default
serde2.rs:9 struct V;
            ^~~~~~~~~
```

As best as I can tell, Rust is able to infer that there's only one implementation possible and uses it, but this information isn't conveyed to the lint, so it marks the struct as being unused.

**Comment by TimNN** [comment]
Adjusting the code for current rust, this no longer produces any warnings (although I did have to use `V` explicitly at one place, but I don't think this can be avoided):

```rust
trait Visitor<R> {
    fn foo();
}

fn visit<V: Visitor<R>, R>() {
    V::foo()
}

struct V;

impl Visitor<()> for V {
    fn foo() { println!("hi") }
}

fn main() {
    visit::<V, ()>()
}
```

**Comment by apasel422** [comment]
I'm not sure that @TimNN's updated code exhibits the original issue, because it explicitly uses `V`.

**Comment by TimNN** [comment]
I was thinking about that when I adapted the code. Do you know a way to avoid explicitly mentioning V?

**Comment by nagisa** [comment]
From the duplicate issue:

```rust
struct A;
trait As {
    fn banana();
}

impl As for A {
    fn banana() { panic!("banana"); }
}

trait Bs<T: As> {
    fn do_it(&self) { T::banana() }
}

struct B;

impl Bs<A> for B {}

fn main() {
    Bs::do_it(&B)
}
```

When compiled, the dead_code lint fires, claiming that `A` is not used.

This code, when [run](https://play.rust-lang.org/?gist=23a48c7157f276c57e2035d9540d1ab9&version=stable&backtrace=0), panics with `banana`, thus proving that `A` is, in fact, used.

---

Removing E-needstest, because the issue is not resolved.

**Comment by Mark-Simulacrum** [comment]
Closing in favor of https://github.com/rust-lang/rust/issues/18290.
